website/blog/2025-07-01-mlflow-prompt-evaluate/index.md

```diff
@@ -0,0 +1,435 @@
+        {
+            "id": 0,
+            "text": "Registration No.",
+            "box": [94,169,191,186],
+            "linking": [
+                [0,1]
+            ],
+            "label": "question",
+            "words": [
+                {
+                    "text": "Registration",
+                    "box": [94,169,168,186]
+                },
+                {
+                    "text": "No.",
+                    "box": [170,169,191,183]
+                }
+            ]
+        },
+        {
+            "id": 1,
+            "text": "533",
+            "box": [209,169,236,182],
+            "label": "answer",
+            "words": [
+                {
+                    "box": [209,169,236,182
+                    ],
+                    "text": "533"
+                }
+            ],
+            "linking": [
+                [0,1]
+            ]
+        }
+    ]
+    }
+```
+
+#### 1. Observe Data
+
+Let's try to read a random annotated file.
+
+```python
+DATA_DIRECTORY = "./data" # for simplicity, a local directory
+ANNOTATIONS_DIRECTORY = os.path.join(DATA_DIRECTORY, "annotations")
+
+def _extract_qa_pairs(data: dict) -> dict:
+    """Extracts question-answer pairs from OCR-style linked data."""
+    qa_pairs = {}
+
+    elements = {item["id"]: item for item in data}
+    for item in data:
+        if item["label"] != "question" or not item["linking"]:
+            continue
+        for q_id, a_id in item["linking"]:
+            if q_id != item["id"]:
+                continue
+            answer = elements.get(a_id)
+            if answer and answer["label"] == "answer":
+                q_text = " ".join(w["text"] for w in item["words"])
+                a_text = " ".join(w["text"] for w in answer["words"])
+                qa_pairs[q_text] = a_text
+    return qa_pairs
+
+def get_json(file_name: str, directory: str = ANNOTATIONS_DIRECTORY) -> dict:
+    file_name += ".json" if not file_name.endswith(".json") else file_name
+    path = os.path.join(directory, file_name)
+    with open(path, "r", encoding="utf-8") as f:
+        contents = json.load(f)["form"]
+        if isinstance(contents, list):
+            if all(isinstance(page, list) for page in contents):
+                flat_items = [item for page in contents for item in page]
+            else:
+                flat_items = contents
+            return _extract_qa_pairs(flat_items)
+        else:
+            return _extract_qa_pairs(contents)
+
+```
+
+What does _\_extract_qa_pairs_ do here? Let's break down it down for a second:
+
+- We create a look-up dictionary _qa_pairs_
+- We identify "question" items that have linked answer pair(s) in the form of _(q_id, a_id)_
+- Construct question-answer pairs by joining the individual words  
+  <br/>
+  Subsequently, we can call _get_json_ to get the OCR-identified question/answer structure based on the form image data.
+
+```python
+
+from utils import get_image, get_json, get_random_files
```

Comment:
Note that they have to save the above code to a utils dir. I think it's easier to omit the import and assume this one large script